### PR TITLE
feat: Add a preview Storage Client

### PR DESCRIPTION
**Pull request by nand4011**
Update the protos to the latest version with store functions.

Add a Preview Storage client with control methods and set/get methods.

### CODE REVIEW

File: sdk/src/storage/config/configuration.rs [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+///
+/// Static, versioned configurations are provided for different environments:
+/// ```
+/// use momento::storage::configurations;
+///
+/// /// Use laptop for local development
+/// let developer_config = configurations::Laptop::latest();
+///
+/// /// Use in_region for a typical server environment
+/// let server_config = configurations::Laptop::latest();
```
Comment (anitarua) [inline sdk/src/storage/config/configuration.rs:15]: don't think there's an in-region config

File: sdk/src/storage/messages/store_value.rs [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+                message: "item is not a utf-8 string".to_string(),
+                error_code: MomentoErrorCode::TypeError,
+                inner_error: None,
+                details: None,
+            }),
+        }
+    }
+}
+
+impl TryFrom<StoreValue> for i64 {
```
Comment (anitarua) [inline sdk/src/storage/messages/store_value.rs:150]: missing one for i32?

File: sdk/src/storage/preview_storage_client.rs [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+use crate::storage::messages::data::get::{GetRequest, GetResponse};
+use crate::storage::messages::momento_store_request::MomentoStorageRequest;
+use crate::storage::messages::store_value::StoreValue;
+use crate::storage::preview_storage_client_builder::{
+    NeedsConfiguration, PreviewStorageClientBuilder,
+};
+use crate::storage::{Configuration, DeleteRequest, DeleteResponse, SetRequest, SetResponse};
+use crate::MomentoResult;
+
+/// Client to work with Momento Store.
```
Comment (anitarua) [inline sdk/src/storage/preview_storage_client.rs:19]:
```suggestion
/// Client to work with Momento Storage.
```

File: sdk/src/storage/preview_storage_client.rs [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+/// ```
+#[derive(Clone, Debug)]
+pub struct PreviewStorageClient {
+    pub(crate) storage_client: StoreClient<InterceptedService<Channel, HeaderInterceptor>>,
+    pub(crate) control_client: ScsControlClient<InterceptedService<Channel, HeaderInterceptor>>,
+    pub(crate) configuration: Configuration,
+}
+
+impl PreviewStorageClient {
+    /// Constructs a PreviewStorageClient to use Momento Store.
```
Comment (anitarua) [inline sdk/src/storage/preview_storage_client.rs:53]:
```suggestion
    /// Constructs a PreviewStorageClient to use Momento Storage.
```

File: sdk/src/storage/messages/data/get.rs [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+/// this is what you're after:
+/// ```
+/// # use momento::storage::GetResponse;
+/// # use momento::MomentoResult;
+/// # let get_response = GetResponse::Hit { value: "value".into() };
+/// use std::convert::TryInto;
+/// let item: MomentoResult<String> = get_response.try_into();
+/// ```
+///
+/// You can also go convert into a `Vec<u8>`, `i64`, or `f64` depending on the type you stored:
```
Comment (anitarua) [inline sdk/src/storage/messages/data/get.rs:113]: is i32 not a possible type?

File: sdk/tests/storage/control.rs [hunk 6]
```diff
@@ -0,0 +1,27 @@
+use momento::storage::CreateStoreResponse;
+use momento::MomentoErrorCode;
+use momento::MomentoResult;
+use momento_test_util::unique_store_name;
+use momento_test_util::CACHE_TEST_STATE;
+
+mod create_delete_list_store {
```
Comment (anitarua) [inline sdk/tests/storage/control.rs:7]: could add a small test for create then list stores

**Review by anitarua — COMMENTED** [review]
overall lgtm!

File: sdk/src/storage/preview_storage_client_builder.rs [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+        })
+    }
+}
+
+impl PreviewStorageClientBuilder<ReadyToBuild> {
+    pub fn build(self) -> MomentoResult<PreviewStorageClient> {
+        let agent_value = &utils::user_agent("sdk");
+
+        let data_channel = utils::connect_channel_lazily_configurable(
+            &self.0.credential_provider.cache_endpoint,
```
Comment (anitarua) [inline sdk/src/storage/preview_storage_client_builder.rs:50]: will probably be storage_endpoint, yeah?

File: sdk/src/storage/messages/data/get.rs (hunk 5 above)
Comment (nand4011) [reply sdk/src/storage/messages/data/get.rs:113]: The call can only return an i64 or f64. I added ways to convert an i32 to a StoreValue, but internally it is an i64.

File: sdk/src/storage/preview_storage_client_builder.rs (hunk 7 above)
Comment (nand4011) [reply sdk/src/storage/preview_storage_client_builder.rs:50]: Good catch. I'll update that.

File: sdk/src/storage/preview_storage_client.rs (hunk 3 above)
Comment (nand4011) [reply sdk/src/storage/preview_storage_client.rs:19]: I want to use the product name here. Will it be called Momento Storage or Momento Store?

**Review by cprice404 — COMMENTED** [review]
just clearing this off of y review list for now, please add me again when it's ready for 👀 !

File: sdk/src/credential_provider.rs [hunk 8]
```diff
@@ -291,6 +310,15 @@ mod tests {
             credential_provider.control_endpoint,
             "https://control plane endpoint"
         );
+        assert_eq!(
+            "https://data plane endpoint",
+            credential_provider.token_endpoint
+        );
+
+        assert_eq!(
```
Comment (nand4011) [inline sdk/src/credential_provider.rs:318]: I followed the existing pattern here, but this doesn't seem correct. Are the token and storage endpoints supposed to resolve to the same thing as the cache endpoint?

Comment (cprice404) [reply sdk/src/credential_provider.rs:318]: let's just remove legacy jwt support

File: sdk/src/storage/preview_storage_client.rs (hunk 4 above)
Comment (cprice404) [reply sdk/src/storage/preview_storage_client.rs:55]: 👍

File: sdk/src/storage/preview_storage_client.rs (hunk 3 above)
Comment (cprice404) [reply sdk/src/storage/preview_storage_client.rs]: Storage, Anita is right

File: sdk/src/credential_provider.rs (hunk 8 above)
Comment (nand4011) [reply sdk/src/credential_provider.rs:318]: I'll make a separate pr for that.

Comment (nand4011) [reply sdk/src/credential_provider.rs:318]: https://github.com/momentohq/client-sdk-rust/pull/366

**Review by anitarua — APPROVED** [review]
lgtm!

**Pull request merged**